### ISSUE TITLE
ctc_greedy_decoder inconsistent with ctc_beam_search_decoder

### ISSUE THREAD

**Issue by JuliusKunze** [tensorflow/tensorflow#9550]
The following extract from the the `ctc_beam_search_decoder` documentation seems to be misleading:
"The `ctc_greedy_decoder` is a special case of the `ctc_beam_search_decoder` with `top_paths=1` and `beam_width=1` (but that decoder is faster for this special case)."

Instead, the following results can be observed: 

| Decoding "AA<ctc_blank>AA" using | `merge_repeated=True` | `merge_repeated=False` |
| --- | --- | --- |
| `tf.nn.ctc_beam_search_decoder(top_paths=1, beam_width=1)` | "A"   | "AA" |
| `tf.nn.ctc_greedy_decoder()` | "AA" | "AAAA" |

To reproduce:

```python
import numpy as np
import tensorflow as tf
from unittest import TestCase


class CtcDecodersTest(TestCase):
    def test(self):
        def decode_greedily(beam_search: bool, merge_repeated: bool):
            aa_ctc_blank_aa_logits = tf.constant(np.array([[[1.0, 0.0]], [[1.0, 0.0]], [[0.0, 1.0]],
                                                    [[1.0, 0.0]], [[1.0, 0.0]]], dtype=np.float32))
            sequence_length = tf.constant(np.array([5], dtype=np.int32))

            (decoded_list,), log_probabilities = \
                tf.nn.ctc_beam_search_decoder(inputs=aa_ctc_blank_aa_logits,
                                              sequence_length=sequence_length,
                                              merge_repeated=merge_repeated,
                                              beam_width=1) \
                    if beam_search else \
                    tf.nn.ctc_greedy_decoder(inputs=aa_ctc_blank_aa_logits,
                                             sequence_length=sequence_length,
                                             merge_repeated=merge_repeated)

            return list(tf.Session().run(tf.sparse_tensor_to_dense(decoded_list)[0]))

        self.assertEqual([0], decode_greedily(beam_search=True, merge_repeated=True))
        self.assertEqual([0, 0], decode_greedily(beam_search=True, merge_repeated=False))
        self.assertEqual([0, 0], decode_greedily(beam_search=False, merge_repeated=True))
        self.assertEqual([0, 0, 0, 0], decode_greedily(beam_search=False, merge_repeated=False))
```

This is confusing and probably not intended.

How to solve this:

- Adapt the documentation or
- Adapt the `ctc_beam_search_decoder` implementation to that of `ctc_greedy_decoder` or vice versa. Both directions would cover my use case ("AA"), this decision would depend on which of the other behaviors ("A" or "AAAA") is needed and which of them could be dropped.

### System information
OSX 12.4, TensorFlow 1.1.0 CPU from binary

**Comment by stefbraun** [comment]
I stumbled upon this today, too. Why are the results not fully equivalent when ctc_beam_search_decoder uses top_paths=1 and beam_size=1?

**Comment by gunan** [comment]
Ping!
@ebrevdo any comments?
I know there are a lot of issues assigned to you, so could you reassign this one who may have time to investigate this?

**Comment by haotianteng** [comment]
Also encounter this problem as well.
And test with this data set:
Logits: [ [[200.0,1.0]] , [[-100.0,200.0]] , [[200.0,1.0]] ]
Which should give 'AA',
The greedy decoder is correct, but the beam search decoder only give 'A'

And the beam search deocder work on this dataset:
[ [[log(0.4),log(0.6)]] , [[log(0.4),log(0.6)]]  ]
Greedy decoder gives an empty output, where the beam search decoder gives the true output  'A'.
Which is known would fail in best path greedy decoding.

But they both fail on this dataset:
[ [[log(0.4),log(0.6)]] , [[log(0.4),log(0.6)]], [[-200.0,200.0]] , [[200.0,-200.0]] ]
Where the true output is 'AA', but both decoders give 'A'.
But I guess they fail in the different place, beam_search decoder seems unable to give the correct answer for 'AbA'-type path, where greedy decoder failed in the first two logits.


To reproduce:
```
logits = tf.Variable(initial_value = [[[np.log(0.4),np.log(0.6)]],[[np.log(0.4),np.log(0.6)]],[[-200.0,200.0]],[[200.0,-200.0]]])
output= tf.nn.ctc_beam_search_decoder(inputs = logits,sequence_length=[4],top_paths = 3,beam_width = 500)
greedy_output = tf.nn.ctc_greedy_decoder(inputs = logits,sequence_length = [4])
sess = tf.Session()
sess.run(tf.global_variables_initializer())
decode_val = sess.run(output)
greedy_val = sess.run(greedy_output)
print decode_val,'\n',greedy_val
```

##System information
Ubuntu 14.04, Tensorflow 1.3.0 CPU from pip install.

**Comment by liufeng900204** [comment]
same issue and look forward to solving that

**Comment by marcoleewow** [comment]
any updates? having the same issue as well.

**Comment by ryanleary** [comment]
I spent a good fraction of the day debugging this before coming to a somewhat obvious realization. If you read the Graves paper, CTC beam decoding implicitly collapses repeated characters as part of calculating the optimal path (i.e. 'AAA' will contribute probability mass through the path 'A'). 

So the correct CTC decoding behavior occurs when `merge_repeated=False`. In this case, it DOES merge repeated characters. The `merge_repeated` flag, when true, will merge repeated characters after characters have already been merged/blank symbols removed. This parameter should simply be removed. As it stands now, the behavior is _extremely_ misleading.

I'm happy to provide a PR for this change if it's something a tensorflower would be amenable with. cc @ebrevdo @gunan

**Comment by marcoleewow** [comment]
Let's use the example from above: `"AA<ctc_blank>AA"`

Expected behavior:

when `merge_repeated=False`, `"AA<ctc_blank>AA"` should give `AAAA`
when `merge_repeated=True`, `"AA<ctc_blank>AA"` should give `AA`

Only `tf.nn.ctc_greedy_decoder()` is working as expected, but `tf.nn.ctc_beam_search_decoder(top_paths=1, beam_width=1)` is not. 

@ryanleary are you saying `tf.nn.ctc_beam_search_decoder(top_paths=1, beam_width=1)` merged symbols twice? I think I might be able to help out in amending the decoder.

**Comment by ryanleary** [comment]
I agree that `merge_repeated=True` should give `AA` is the correct behavior. What I'm saying is that having a case where `merge_repeated=False` in a beam decoder doesn't make a whole lot of sense. Further, the merging is somewhat fundamental to the way the algorithm handles summing probabilities across multiple paths.

In the current implementation of the beam decoder, if `merge_repeated=False`, repeated characters are merged once. If `merge_repeated=True`, they are merged *again*, returning nonsensical results.

I'm advocating for removing the flag, and the *only* behavior to be where repeated characters are merged (but equivalent to the `merge_repeated=False` current functionality).

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by ebrevdo** [comment]
following up on PR

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 141 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 156 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 171 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 186 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 201 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 216 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ebrevdo** [comment]
Since this is a backwards breaking API change, I'll remove the argument for TF 2.0.  Thank you all for the report & detailed analysis.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @ebrevdo: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.
